File: CLASS 6/1086.c

// 1086번 - 박성원

/* Include */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* Define */
#define MAX(a,b) a>b?a:b
#define MIN(a,b) a>b?b:a
#define f(i,N) for(int i=0; i<N; i++)

#define LENGTH 50
#define DIGITS 100

/* Struct */

/* Global Var */
int** dp; // 50(최대 50 자리) * 10(0~9)
int divisor; // 유일한 나누기 수

int* divisors; // 각 배열의 정리된 나머지
int* lengths;  // 각 배열의 길이

int N;

/* Function Declarations */

/* Main */
int main(void){
    dp = (int**)malloc(LENGTH * sizeof(int*));
    f(i,LENGTH){
        dp[i] = (int*)malloc(DIGITS * sizeof(int));
        f(j,DIGITS)
            dp[i][j] = -1;
    }
    char arr[51]; // 입력받을 문자열
    scanf("%d",&N);
    divisors = (int*)malloc(N*sizeof(int));
    lengths = (int*)malloc(N*sizeof(int));
    f(i,N){
        scanf("%s",arr);
        int dig = strlen(arr)-1;
        for(int j=0; arr[j] != 0; j++){
            divisors[j] += getRemainder(dig, arr[j] - '0');
            dig--;
        }
    }

    return 0;
}

/* Functions */
int getRemainder(int length, int digit){
    if(dp[length][digit] != -1)
        return dp[length][digit];
    if(length == 0)
        return digit;
    return dp[length][digit] = getRemainder(length-1, (digit*10) % divisor);
}

File: CLASS 6/1086.py
# 1086번 - 박성원

# ---------- import 선언부 ----------


# ---------- 전역변수 선언부 ----------


# ---------- 클래스 선언부 ----------

    
# ---------- 함수 선언부 ----------

def get_remainder(length, digit, divisor, dp):
    if dp[length][digit] != -1:
        return dp[length][digit]
    if length == 0:
        return digit
    dp[length][digit] = get_remainder(length - 1, (digit * 10) % divisor, divisor, dp)
    return dp[length][digit]


# 여기서부터 메인 코드가 시작됨

N = int(input())
divisor = int(input())  # divisor를 입력받음
dp = [[-1] * 10 for _ in range(50)]  # DP 테이블 초기화 (길이 50, 숫자 0-9)

divisors = [0] * N
lengths = [0] * N

for i in range(N):
    arr = input().strip()
    dig = len(arr) - 1
    for j in range(len(arr)):
        divisors[i] += get_remainder(dig, int(arr[j]), divisor, dp)
        dig -= 1